lib/libmbedtls/core/cmac.c

```diff
@@ -0,0 +1,253 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2018, ARM Limited
+ * Copyright (C) 2019, Linaro Limited
+ * Copyright (C) 2021, SumUp Services GmbH
+ */
+
+#include <assert.h>
+#include <crypto/crypto.h>
+#include <crypto/crypto_impl.h>
+#include <kernel/panic.h>
+#include <mbedtls/cipher.h>
+#include <mbedtls/cmac.h>
+#include <stdlib.h>
+#include <string.h>
+#include <tee_api_types.h>
+#include <utee_defines.h>
+#include <util.h>
+
+struct mbed_cmac_ctx {
+	struct crypto_mac_ctx mac_ctx;
+	mbedtls_cipher_context_t cipher_ctx;
+};
+
+static const struct crypto_mac_ops mbed_des3_cmac_ops;
+static const struct crypto_mac_ops mbed_aes_cmac_ops;
+
+static struct mbed_cmac_ctx *to_cmac_ctx(struct crypto_mac_ctx *ctx, const mbedtls_cipher_id_t cipher_id)
+{
+	assert(ctx);
+
+	if (cipher_id == MBEDTLS_CIPHER_ID_AES)
+		assert(ctx->ops == &mbed_aes_cmac_ops);
+
+	if (cipher_id == MBEDTLS_CIPHER_ID_3DES)
+		assert(ctx->ops == &mbed_des3_cmac_ops);
+
+	return container_of(ctx, struct mbed_cmac_ctx, mac_ctx);
+}
+
+static TEE_Result mbed_cmac_init(struct crypto_mac_ctx *ctx,
+		const uint8_t *key, size_t len, const mbedtls_cipher_id_t cipher_id)
+{
+	if (cipher_id != MBEDTLS_CIPHER_ID_3DES && cipher_id != MBEDTLS_CIPHER_ID_AES)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	struct mbed_cmac_ctx *c = to_cmac_ctx(ctx, cipher_id);
+
+	const mbedtls_cipher_info_t *cipher_info = NULL;
+
+	cipher_info = mbedtls_cipher_info_from_values(cipher_id,
+			len * 8,
+			MBEDTLS_MODE_ECB);
```

Comment:
Indentation

Reply:
minor: indentation still to fix (indent to opening parenthesis)